### PR TITLE
Return error when trying to use prediction resistance

### PR DESCRIPTION
**Pull request by kroeckx**

### CODE REVIEW

**Comment by kroeckx** [comment]
I do remember that discussion, but the link doesn't seem to work and I can't find it #4402.

It might be that some of the entropy sources do qualify as a NRGB, but in my opinion none of default sources, like a parent DRBG, /dev/urandom, getentropy() qualify in my understanding. Something like RDSEED might qualify, while RDRAND not.

Anyway, since we don't seem to agree on this, maybe we should ask others? Maybe people from NIST?

**Comment by kroeckx** [comment]
I've moved to check to just before we get the system entropy.

**Comment by mspncp** [comment]
> I do remember that discussion, but the link doesn't seem to work and I can't find it #4402.

The link did not work on Chrome (but it worked with Internet Explorer) because it's inside a review for an outdated commit.  You'll see it if you press 'Show outdated' on your second review from  22 Oct 2017.

**Comment by kroeckx** [comment]
Both the current and previous patch actually allow you to use prediction resistance, just not with the default methods to get entropy.

**Comment by mspncp** [comment]
The use of the prediction resistance parameter is that you can say: „reseed immediately,  before generating the random output, because I need extra security!“ And your proposal is that the random generator should cowardly fail because it is afraid that its entropy source is not good enough. That does not make sense to me.

**Comment by paulidale** [comment]
I'm not seeing the point of this either.

**Comment by mspncp** [comment]
The DRBG is the determinstic part of a hybrid random generator and as such, the quality of its output depends on the quality of its random source. But it is not the task of the building block to decide whether the random source is good enough or not. The choice of the proper entropy source could be mandated by a future FIPS certification. But for now, it is consensus that the os random generator is not that bad after all. And requesting prediction resistance should be possible regardless of the choice of random source. So we should leave this question for a future FIPS evaluation.

**Comment by kroeckx** [comment]
So my understanding is that you're saying the kernel generated data is (currently) good enough. But that seems to mean to me that we need to either call the parent with the prediction_resistance parameter, or ask the kernel ourselves. So should I just remove that if that returns an error?

**Comment by mspncp** [comment]
Sorry Kurt, I did not have much time this weekend to answer your question. You will receive an answer by tomorrow or the day after.

**Comment by kroeckx** [comment]
@mspncp: Ping

**Comment by mspncp** [comment]
First of all, I would like to repost a minimally redacted version of my post from 22 Oct 2017, since the original is not easily available due to a bug in the GitHub webinterface (the link only works with IE, not with Chrome).


## mspncp on 25 Oct 2017

Currently the `prediction_resistance` request is not propagated to the parent. So a generate request to the private DRBG with `prediction_resistance=1` only reseeds the private DRBG, not the master DRBG.


It is my understanding that the NIST standard does not mandate the prediction request to be propagated to the parent DRBG: The NIST standard only considers a single DRBG. An adversary might have gained knowledge about its internal state, but not about its parent, which is considered a black box. Here are some citations from [NIST SP 800-90Ar1](http://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf):

*Definition of Prediction Resistance (page 6)*

> An RBG provides prediction resistance relative to time T if it provides assurance that an adversary with knowledge of the state of the RBG at some time(s) prior to T (but incapable of performing work that matches the claimed security strength of the RBG) would be unable to distinguish between observations of ideal random bitstrings and (previously unseen) bitstrings output by the RBG at or subsequent to time T. In particular, an RBG whose design allows the adversary to step forward from the initially compromised RBG state(s) to obtain knowledge of subsequent RBG states and the corresponding outputs (including the RBG state and output at time T) would not provide prediction resistance relative to time T.

On page 25, it says 

> The prediction_resistance_request parameter indicates whether or not prediction resistance is to be provided during the request (i.e., whether fresh entropy is required[4]).

Chaining of DRBGs is justified by footnote[4] of page 25:

> [4] 4 Entropy input may be obtained from an entropy source or an NRBG, both of which provide fresh entropy. Entropy input could also be obtained from a DRBG that has access to an entropy source or NRBG. The request for prediction resistance rules out the use of a DRBG that does not have access to either an entropy source or NRBG.

and in further detail by footnote [5], page 30:
> A DRBG may be reseeded by an entropy source or an NRBG, both of which provide fresh entropy. A DRBG could also be reseeded by a DRBG that has access to an entropy source or NRBG. The request for prediction resistance during reseeding rules out the use of a DRBG that does not have access to either an entropy source or NRBG. See [SP 800-90C] for further discussion. 

In our case one might argue differently, because the master DRBG resides in the same address space as public and private DRBG. An adversary that gains knowledge about the private DRBG could indeed also have knowledge about its parent. I tried to mitigate this possility by lowering the reseed intervals to keep the total "os reseed interval" at 2^24. But I would understand if someone insisted on forwarding the prediction_resistance request to the parent. In this case one would indeed get the Domino effect you mentioned. It's the usual security vs. performance tradeoff. As such, it's a political question and should probably be decided the OMC members. I am a little in favor of my choice, but would also follow a majority vote.



## kroeckx on 22 Oct 2017

I have no idea how you can interpret that text differently. There needs to be a way to get fresh entropy from somewhere. A DRBG does not provide fresh entropy, you need something that directly talks to the entropy source to get such fresh entropy and that's the master DRBG.

## mspncp on 25 Oct 2017

I disagree, and it is not a matter of interpretation, it is written down in the standard:

- Requesting prediction resistance simply means requesting a reseeding of the DRBG before generating random output. There is no difference between a "normal" reseeding and a reseeding triggered by a prediction resistance request.
- The NIST standard explicitely allows reseeding from a DRBG which is connected to an approved entropy source. There is no word about a need to reseed the parent immediately before obtaining its random output.

If you are questioning the fact that a DRBG can safely be reseeded from the master DRBG, then you question the entire construction of chaining the DRBGs to improve the security of the DRBGs in the first place.

But before going deeper into this, I would like to dispel your concerns about prediction resistance:

I agree that a kernel mode CSPRNG is more secured than a user space CSPRNG. So I completely understand if someone prefers to pull randomness from the former instead of the latter. But this is already possible to do: If the applications needs "strong" prediction resistance for the public or the private DRBG, all it has to do is to call RAND_poll() before calling RAND_bytes() resp. RAND_priv_bytes(). This will reseed the master DRBG from the os entropy sources, and the reseeding will then be propagated to the public resp. private DRBG. Alternatively, it could obtain a pointer to the master DRBG (drbg = RAND_DRBG_get0_master()) and call its generate method requesting prediction resistance.


### The facts

I was not trying to interpret the standard, I was stating the facts from the document and verifying them by giving quotations. So please tell me, where exactly I misinterpreted something in my reasoning below. And please point out the reference where it is stated that the parent DRBG's generate method SHALL be called demanding prediction resistance.

The DRBG construction is an approved method of 'stretching' the sparse resource of randomness into a longer pseudo random output. This means that the DRBG's output is computationally indistinguishable from a true random coin toss for an adversary whose computational capabilities lie within the limits assumed for the given security level (not more and not less). This holds under the assumptions that

- the DRBG was reseeded from an approved entropy source or nondeterministic RNG
- the internal state of the DRBG is unknown to the adversary

If the previous statement holds then not only the first block of 128 bits generated by the parent DRBG is computationally indistinguishable from true random output, but also the second and the third block, etc. So why should the first 128 bit block contain "more entropy" than the second? Neither of them contains any entropy because the output is a deterministic function of the internal state. All the randomness lies in the random choice of this internal state, and the fact that the internal state remains secret.

Nothing else happens when you reseed from the kernel's RNG: You will get some block of pseudo random output from a deterministic RNG and you can't even be certain that enough bits of "fresh" entropy have been added to the kernel's random pool since your last request (more on this subject below).

Calling RAND_DRBG_generate() with prediction_resistance=1 is the request to "reseed before generating output", in order to handle the case that the internal state of the DRBG might have been (partially) compromised. It triggers a "normal" reseed, without any additional constraints on the "quality" of the random input pulled from the entropy source.

The NIST standard explicitely states that instead of an approved nondeterministic RNG or entropy source it is allowed to use a DRBG which is seeded by an approved nondeterministic RNG or entropy source. Indeed, unless the state of the parent DRBG is compromised, too, or the randomness input fed into the child DRBG is intercepted by the adversary, this reseeding will immediately restore the integrity of the child DRBG.


### The myth

What is "fresh" entropy in the first place? Is it filled into the DRBG and used up by generate requests like the gasoline of an electricity generator (pun intended)? No, this is just an analogy for humans and taking this analogy too literal is the cause for the misbelief that a DRBG which has been seeded with, say, 128 bits of entropy gets drained after generating 128 bits of random output. This superstition caused generations of programmers to block on /dev/random [waiting for "fresh" entropy](https://www.2uo.de/myths-about-urandom/#low-entropy). I don't know the author of the linked article, but I agree with his arguments, in particular with the [quote of Daniel Bernstein](https://www.2uo.de/myths-about-urandom/#experts), which was taken from [this email](https://www.mail-archive.com/cryptography@randombit.net/msg04763.html)).


### Theory vs. practice

In the context of the NIST document the case is clear: it is not necessary to propagate the prediction resistance request to the master. In our case the situation is a little more complicated, because both DRBGs are located in the address space of the same process, so it is not clear whether the master DRBG's internal state can be considered secure after the child DRBG's internal state was compromised.

Whether or not the master DRBG can be considered secure depends on the threat model. If the adversary can only intercept network traffic and learn about the state of the child DRBG, but does not get to see any of the master DRBG's output, then the answer is clearly yes. (Apropos: it's probably better not to reseed the ssl DRBGs from the public DRBG, because we have no control how much output of the public DRBG will be used for other purposes and become publicly known.)

If the adversary obtained knowledge of the child DRBG's internal state by reading some portion of the process memory, then it is arguable that he might know about the internal state of the master, too. One mitigation for this threat would be to not use static variables to hold the internal state of the three DRBGs, but allocate it from the secure heap.

Anyway, as stated above: It is easy to trigger a full reseed from the os entropy sources either by calling RAND_poll() or by using a DRBG which is not chained and requesting prediction resistance.

**Comment by mspncp** [comment]
#### TL;DR: Propagating prediction resistance requests is not mandated by the standard, it's a new feature

1) From the NIST standard's point of view, a prediction resistance request is equivalent to a request to reseed immediately before generating the bytes. Not more, not less. 
2) Your proposal to propagate the prediction resistance request to the parent is a deviation from the NIST standard, let's call it an additional feature.

Now it is legitimate to ask the question whether this additional feature increases security? My answer is

#### This feature does not add any new security that could not be achieved using the existing features. 

For example, if I were paranoid, I could set the `reseed_interval` of `<master>` and `<private>` to 1. Then every call to `RAND_priv_bytes()` would trigger a full reseed of both shared DRBGs. But note that there is one more side effect: A reseeding of `<master>` will trigger a reseeding of `<public>`, because we also implemented downward propagation from `<master>` to its children. This is also an additional feature that deviates from the NIST standard, but it was necessary to do it in order to fix the `RAND_add()`/`RAND_bytes()` pattern. Your upward propagation of prediction resistance requests only adds additional complexity without adding security.

If I were even more paranoid and wouldn't trust DRBG chaining, then I could use my private unchained DRBG instance and either set the reseed_interval to 1, or specify prediction_resistance=1 on every generate request.

That said, one final remark: The RAND API does not support requesting prediction resistance. Nowhere in libcrypto is it used currently. So this pull request adds additional complexity for the empty set.

#### Summary

-1 from me: This pull request should be closed.

**Comment by kroeckx** [comment]
I think the main point of the prediction_resistance_request parameter is to try to recover from the a possible compromise, and you do that by reseeding. If says that you need "fresh entropy" for that.

But then when I actually try to write down how you would implement fresh entropy, I can only come up with 2 definitions:
- All the bits collected by the source need to be collected after the last generate call
- It can have been collected before the last generate call, we just didn't use it yet. And this mostly depends on the next topic.

In the end it's about which of the DRBGs in the chain have been compromised, and I guess there are several separation levels:
- The DRBG itself
- The process
- The kernel
- The hardware

I guess that if the process itself is compromised, there is little point in trying to recover.

**Comment by kroeckx** [comment]
(edited previous reply slightly)

**Comment by kroeckx** [comment]
Since you don't know which DRBGs all have been compromised, it seems to me that NIST standard have resolved this by requiring all DRBGs in the chain to reseed. This is based on that the get_entropy and reseed functions have a prediction_resistance_request parameter indicating that it's not a normal one, it needs to get fresh entropy. But I think what it really wants is non-compromised entropy, not fresh entropy.

**Comment by kroeckx** [comment]
I've been reading SP 800-90C again, and in 10.1.1 it says that if get_entropy was called with prediction resistance that the generate call should be called with prediction resistance.

**Comment by mspncp** [comment]
I skimmed over SP 800-90C this evening, too. I still need some time to read the document more thoroughly, but it seems now that you are probably right and that I was too much focused on SP 8000-90A and did not consider the changed requirements in SP 800-90C. If my last post sounded too harsh, then I apologize for it.

In view of the approaching deadline for the beta freeze we should probably change the signature of the get_entropy callback before publishing the API, but otherwise postpone the discussions about prediction resistance until after the freeze, because it will only become relevant for the upcoming FIPS evaluation. I need to concentrate on finishing the documentation now, otherwise it is possible that we will end up with no public RAND_DRBG API at all for 1.1.1.

File: crypto/rand/rand_lib.c [hunk 1]
```diff
@@ -216,6 +216,12 @@ size_t rand_drbg_get_entropy(RAND_DRBG *drbg,
         }
 
     } else {
+       if (prediction_resistance) {
```
Comment (mspncp) [inline crypto/rand/rand_lib.c]: indentation is only 7 spaces instead of 8

File: crypto/rand/rand_lib.c [hunk 2]
```diff
@@ -216,6 +216,12 @@ size_t rand_drbg_get_entropy(RAND_DRBG *drbg,
...
+            /* We don't have any entropy sources that comply with the NIST standard
+            * to provide prediction resistance. */
+            return 0;
```
Comment (mspncp) [inline crypto/rand/rand_lib.c]: Well, sigh, after reading NIST SP 800-90C I have to admit that you are right. The notion of prediction resistance is formulated much more precisely there and much stricter than in  NIST SP 800-90C. Since we can't fullfill the requirements, it is consequent to fail.

However, the function should generate an error message for the caller like that:
```
+            /* We don't have any entropy sources that comply with the NIST standard
+            * to provide prediction resistance. */
+            RANDerr(RAND_DRBG_GET_ENTROPY,·PREDICTION_RESISTANCE_NOT_SUPPORTED);
+            return 0;
```

Comment (mspncp) [reply crypto/rand/rand_lib.c]: Or even better, with reference (please also in commit message).
```
+            /* We don't have any entropy sources that comply with the NIST standard
+             * to provide prediction resistance (see NIST SP 800-90C, Section 5.4).
+             */
+            RANDerr(RAND_DRBG_GET_ENTROPY,·PREDICTION_RESISTANCE_NOT_SUPPORTED);
+            return 0;
```
I will take care of documenting the new `get_entropy` callback signature and add a statement to the manpage that prediction resistance is currently not supported yet.

**Comment by kroeckx** [comment]
New commit pushed

File: crypto/rand/drbg_lib.c [hunk 3]
```diff
@@ -377,7 +379,9 @@ int RAND_DRBG_reseed(RAND_DRBG *drbg,
     drbg->state = DRBG_ERROR;
     if (drbg->get_entropy != NULL)
         entropylen = drbg->get_entropy(drbg, &entropy, drbg->strength,
-                                   drbg->min_entropylen, drbg->max_entropylen);
+                                      drbg->min_entropylen,
+                                      drbg->max_entropylen,
+                                      prediction_resistance);
```
Comment (mspncp) [inline crypto/rand/drbg_lib.c]: My indentation was even worse, but yours is still off-by-one, you need to add one more space (lines 382-384).

File: crypto/rand/rand_lib.c [hunk 4]
```diff
@@ -172,7 +172,8 @@ size_t rand_acquire_entropy_from_cpu(RAND_POOL *pool)
  */
 size_t rand_drbg_get_entropy(RAND_DRBG *drbg,
                         unsigned char **pout,
-                        int entropy, size_t min_len, size_t max_len)
+                        int entropy, size_t min_len, size_t max_len,
+                        int prediction_resistance)
```
Comment (mspncp) [inline crypto/rand/rand_lib.c]: Lines 174-176 are not properly indented (not your fault).

File: crypto/rand/rand_lib.c [hunk 5]
```diff
@@ -225,6 +226,17 @@ size_t rand_drbg_get_entropy(RAND_DRBG *drbg,
...
 
     } else {
+        if (prediction_resistance) {
+            /*
+             * We don't have any entropy sources that comply with the NIST
+             * standard to provide prediction resistance (see NIST SP 800-90C,
+             * Section 5.4).
+             */
+            RANDerr(RAND_F_RAND_DRBG_GET_ENTROPY,
+                RAND_R_PREDICTION_RESISTANCE_NOT_SUPPORTED);
```
Comment (mspncp) [inline crypto/rand/rand_lib.c]: Indentation
